### PR TITLE
Slideshow virtualization

### PR DESCRIPTION
**Pull request by shmuelhizmi**
what - virtualizing the slideshow to only load images in the view or near it (configurable)
why - we need this functionality for the fullscreen view to prevent all images and videos loading in full resolution right when the fullscreen opens

### CODE REVIEW

**Review by nirnaor — CHANGES_REQUESTED** [review]
1. Description of what your PR is doing
2. Tests for new logic
3. No formatting changes in PRS

**Review by yhattav — CHANGES_REQUESTED** [review]
remove lint changes

File: packages/gallery/package.json [hunk 1]
```diff
@@ -33,6 +33,7 @@
     "test:e2e-styleParams": "jest --testPathPattern='.*/e2e/styleParams/.*.e2e.spec.js'",
     "test:e2e-integration": "jest --testPathPattern='.*/e2e/integration/.*.e2e.spec.js'",
     "test:e2e-ssr": "jest --setupFiles=./tests/ssr-config.js --testPathPattern='.*/e2e/ssr/.*.e2e.spec.js'",
+    "test:spec": "jest --testPathPattern='.*/components/.*.spec.js'",
```
Comment (nirnaor) [inline packages/gallery/package.json]: remove

File: packages/gallery/src/components/gallery/proGallery/slideshowView.js [hunk 2]
```diff
@@ -984,8 +984,52 @@ class SlideshowView extends React.Component {
     ];
   }
 
+  get isScrollable() {
+    return this.props.options.slideAnimation ===
+  GALLERY_CONSTS.slideAnimations.SCROLL;
+  }
+
+  getBufferedItems(galleryGroups, container) {
+    const groups = this.props.getVisibleItems(galleryGroups, container);
```
Comment (nirnaor) [inline packages/gallery/src/components/gallery/proGallery/slideshowView.js]: remove `this` dependency

File: packages/gallery/src/components/gallery/proGallery/slideshowView.js [hunk 3]
```diff
@@ -984,8 +984,52 @@ class SlideshowView extends React.Component {
...
+    const { enabled = false, rightItemMargin = 3, leftItemMargin = 3, rightItemScrollMargin = 10, leftItemScrollMargin = 10 } = this.props.options.behaviourParams?.gallery?.horizontal?.itemVirtualization || {};
+    if (!enabled) {
+      return groups.map((group) => ({ group, shouldRender: true }));
+    }
```
Comment (nirnaor) [inline packages/gallery/src/components/gallery/proGallery/slideshowView.js]: remove from component

**Review by nirnaor — CHANGES_REQUESTED** [review]

File: packages/gallery/src/components/gallery/proGallery/slideshowView.js [hunk 4]
```diff
@@ -984,8 +987,24 @@ class SlideshowView extends React.Component {
     ];
   }
 
+
+  getBufferedItems(galleryGroups, container) {
+    const { state, props } = this;
+    const { options } = props;
+    const { activeIndex } = state;
+    const groups = this.props.getVisibleItems(galleryGroups, container);
```
Comment (nirnaor) [inline packages/gallery/src/components/gallery/proGallery/slideshowView.js]: [question]
Why didn't you destruct the `getVisibleItems and galleryContainerRef` from `this.props` like you did with `options`

File: packages/gallery/src/components/gallery/proGallery/slideshowView.js [hunk 5]
```diff
@@ -1013,23 +1032,26 @@ class SlideshowView extends React.Component {
...
+          layoutGroupView.map(({group, shouldRender}) => {
+            return group.rendered
+              ? React.createElement(GroupView, {
+                  allowLoop:
+                    this.props.options.slideshowLoop &&
+                    this.props.galleryStructure.width >
+                      this.props.container.width,
+                  ...group.renderProps(galleryConfig),
+                  ariaHidden: group.idx > this.skipFromSlide,
+                  shouldRenderEmpty: !shouldRender,
```
Comment (nirnaor) [inline packages/gallery/src/components/gallery/proGallery/slideshowView.js:861]: [question]
You could have simply returned only the `shouldRender`ed groups and then you could remove the ternary (`:` operator). What do you think?

File: packages/gallery/src/components/helpers/slideshowVirtualization.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  galleryWidth: number;
+  activeIndex: number;
+}) {
+  const {
+    enabled = false,
+    rightItemMargin = 3,
+    leftItemMargin = 3,
+    rightItemScrollMargin = 10,
+    leftItemScrollMargin = 10,
+  } = options.behaviourParams?.gallery?.horizontal?.itemVirtualization || {};
```
Comment (nirnaor) [inline packages/gallery/src/components/helpers/slideshowVirtualization.ts]: @yhattav is it necessary to check all the `?` stuff? Aren't we assuring new SP will always be present?

File: packages/gallery/src/components/helpers/slideshowVirtualization.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    if (firstIndex > activeIndex) {
+      accoumilatedRightMargin += groupPrecOfScreen;
+      if (accoumilatedRightMargin > rightRenderBuffer) {
+        return { group, shouldRender: false };
+      }
+    }
+    if (lastIndex < activeIndex) {
+      accoumilatedLeftMargin += groupPrecOfScreen;
+      if (accoumilatedLeftMargin > leftRenderBuffer) {
+        return { group, shouldRender: false };
```
Comment (nirnaor) [inline packages/gallery/src/components/helpers/slideshowVirtualization.ts]: Lines 40 - 43 and lines 46- 49 are duplicated. Can you try and reuse?

File: packages/gallery/tests/components/gallery/slideshowView.spec.js [hunk 8]
```diff
@@ -204,6 +208,54 @@ describe('Slideshow View', () => {
...
+          gallery: {
+            horizontal: {
+              itemVirtualization: {
+                enabled: true,
+                rightItemMargin: 0,
+                leftItemMargin: 0,
+              },
+            },
+          },
+        },
```
Comment (nirnaor) [inline packages/gallery/tests/components/gallery/slideshowView.spec.js]: Please don't duplicate tests and change something small. I can't understand the difference between the tests without reading it slowly. The difference between tests should be understood from the code, not from the `it`

File: packages/lib/src/common/utils/index.js [hunk 9]
```diff
@@ -436,6 +436,14 @@ class Utils {
     }
   }
 
+  awaitStateChange(that, state) {
```
Comment (nirnaor) [inline packages/lib/src/common/utils/index.js]: Why did you add a method that no one is using?

**Review by nirnaor — CHANGES_REQUESTED** [review]

File: packages/lib/src/common/utils/index.js [hunk 10]
```diff
@@ -436,6 +436,7 @@ class Utils {
     }
   }
 
+
```
Comment (nirnaor) [inline packages/lib/src/common/utils/index.js:471]: remove

File: packages/gallery/src/components/helpers/virtualization.ts [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+  if (!enabled) {
+    return items.map((item) => ({
+      item,
+      shouldRender: true,
+    }));
+  }
+  const isHorizontal =
+    options.scrollDirection === GALLERY_CONSTS.scrollDirection.HORIZONTAL;
+  const size = isHorizontal ? galleryWidth : galleryHeight;
```
Comment (nirnaor) [inline packages/gallery/src/components/helpers/virtualization.ts:117]: nice

File: packages/lib/src/common/interfaces/virtualization.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  /**
+   * @description enable item virtualization and only load items that are visible on the screen or in the screen margin
+   * @default false
+   */
+  enabled?: boolean;
+  /**
+   * @description how many item margins (item / gallery-screen-size) to load as a buffer to the right
+   * @default 3
+   */
+  forwardItemMargin?: number;
```
Comment (nirnaor) [inline packages/lib/src/common/interfaces/virtualization.ts:11]: [question]
how will your code work when all these items are `null`? Why are they optional

**Review by nirnaor — CHANGES_REQUESTED** [review]
Please attach:

1. Passing canary 
2. Using that canary - url to the site containing the bug solved

**Pull request merged**